modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -480,6 +550,183 @@ public int getZoneId(String name) {
         }
     }
 
+    /**
+     * The method for obtaining data nodes of the specified zone.
+     * If the {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleUp}
+     * or {@link DistributionZoneConfigurationSchema#dataNodesAutoAdjustScaleDown}
+     * equals to 0 then returned completable future will be completed when data nodes will be saved to the meta storage.
+     *
+     * @param zoneId Zone id.
+     * @param topVer Topology version.
+     * @return The data nodes future.
+     */
+    public CompletableFuture<Set<String>> getDataNodes(int zoneId, long topVer) {
```

Comment:
What should we return in this method in scenario, when scale up / down are non-null, and we restart some node? So, scenario is: we had cluster with data nodes [A,B,C], C was restarted, and some table is going to be created from C after restart, which data nodes we should get to calculate assignments? In your code, it seems that we will return an empty data nodes, is it a correct behaviour?

Reply:
Note, that I'm not talking about the first start of the cluster, where we agreed to return an empty data nodes, as far as user set non-immediate scale up/down

Reply:
I tried to fix it. Please look at changes in DistributionZoneManager#initDataNodesFromVaultManager.